File: DSA/Set/javascript/built-in_set.js
#!/usr/bin/env node
/**
 * Module: setExamples
 * Description: Demonstrates built-in JavaScript Set usage with core operations.
 */

/**
 * Demonstrates basic Set operations in JavaScript including creation, add/delete,
 * membership check, clear, and getting size.
 */
function demonstrateSetOperations() {
  // 📌 Creating a Set
  const s = new Set(); // Empty set
  const s2 = new Set([1, 2, 3]); // Set with initial values

  // ✅ Common Methods
  s2.add(4); // Add element
  s2.delete(2); // Remove element
  console.log(s2.has(3)); // Membership check: true/false
  s2.clear(); // Remove all elements
  console.log(s2.size); // Get size of set

  // 🧠 Advanced Operations (manual in JS)
  const a = new Set([1, 2, 3]);
  const b = new Set([2, 3, 4]);

  const union = new Set([...a, ...b]);
  const intersection = new Set([...a].filter((x) => b.has(x)));
  const difference = new Set([...a].filter((x) => !b.has(x)));

  console.log(union); // Set {1, 2, 3, 4}
  console.log(intersection); // Set {2, 3}
  console.log(difference); // Set {1}
}

demonstrateSetOperations();

File: DSA/Set/python/built-in_set.py
#!/usr/bin/env python3
"""
Module: set_examples
Description: Demonstrates built-in Python set usage with core and advanced operations.
"""

def demonstrate_set_operations():
    """
    Demonstrates basic set operations in Python including creation, add/remove,
    membership check, clear, and advanced set algebra.
    """
    # 📌 Creating a Set
    s = set()                # Empty set
    s = {1, 2, 3}            # Set with values

    # ✅ Common Methods
    s.add(4)                 # Add element
    s.remove(2)              # Remove element (throws error if not present)
    s.discard(5)             # Remove safely (no error if not found)
    print(3 in s)            # Membership check: True/False
    s.clear()                # Remove all elements
    print(len(s))            # Get size of set

    # 🧠 Advanced Operations
    a = {1, 2, 3}
    b = {2, 3, 4}

    print(a.union(b))        # {1, 2, 3, 4}
    print(a.intersection(b)) # {2, 3}
    print(a.difference(b))   # {1}


if __name__ == "__main__":
    demonstrate_set_operations()
